cdap-tms/src/main/java/co/cask/cdap/messaging/store/leveldb/LevelDBMessageTable.java

```diff
@@ -0,0 +1,204 @@
+  public CloseableIterator<Entry> fetch(TopicId topicId, MessageId messageId, boolean inclusive, final int limit,
+                                        @Nullable Transaction transaction) throws IOException {
+    byte[] topic = MessagingUtils.toBytes(topicId);
+    byte[] startRow = Bytes.add(topic, Bytes.toBytes(messageId.getPublishTimestamp()),
+                                Bytes.toBytes(messageId.getSequenceId()));
+    if (!inclusive) {
+      startRow = Bytes.incrementBytes(startRow, 1);
+    }
+    byte[] stopKey = Bytes.stopKeyForPrefix(topic);
+    return new LevelDBCloseableIterator(levelDB, startRow, stopKey, limit, transaction);
+  }
+
+  @Override
+  public void store(Iterator<Entry> entries) throws IOException {
+    long writeTs = System.currentTimeMillis();
+    if (writeTs != writeTimestamp) {
+      pSeqId = 0;
+    }
+    writeTimestamp = writeTs;
+    while (entries.hasNext()) {
+      Entry entry = entries.next();
+      byte[] tableKey = Bytes.add(MessagingUtils.toBytes(entry.getTopicId()), Bytes.toBytes(writeTimestamp),
+                                  Bytes.toBytes(pSeqId++));
+      Long txWritePtr = null;
+      byte[] payload = null;
+      if (entry.isPayloadReference()) {
+        txWritePtr = entry.getTransactionWritePointer();
+      } else {
+        payload = entry.getPayload();
+        if (entry.isTransactional()) {
+          txWritePtr = entry.getTransactionWritePointer();
+        }
+      }
+      levelDB.put(tableKey, generateValue(txWritePtr, payload));
+    }
+  }
+
+  @Override
+  public void delete(TopicId topicId, long transactionWritePointer) throws IOException {
+    byte[] targetTxBytes = Bytes.toBytes(transactionWritePointer);
+    byte[] rowPrefix = MessagingUtils.toBytes(topicId);
+    byte[] stopRow = Bytes.stopKeyForPrefix(rowPrefix);
+
+    List<byte[]> rowKeysToDelete = new ArrayList<>();
+    try (CloseableIterator<Map.Entry<byte[], byte[]>> rowIterator = new DBScanIterator(levelDB, rowPrefix, stopRow)) {
+      while (rowIterator.hasNext()) {
+        Map.Entry<byte[], byte[]> candidateRow = rowIterator.next();
+        byte[] rowKey = candidateRow.getKey();
+        Map<String, byte[]> columns = parseValue(candidateRow.getValue());
+        if (columns != null && columns.containsKey(TX_COL)) {
+          byte[] txPtr = columns.get(TX_COL);
+          if (Bytes.equals(txPtr, targetTxBytes)) {
+            rowKeysToDelete.add(rowKey);
+          }
+        }
+      }
+    }
+
+    for (byte[] deleteRowKey : rowKeysToDelete) {
+      levelDB.delete(deleteRowKey);
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    // no-op
+  }
+
+  private static byte[] generateValue(Long txWritePtr, byte[] payload) {
+    Map<String, byte[]> data = new HashMap<>();
+    if (txWritePtr != null) {
+      data.put(TX_COL, Bytes.toBytes(txWritePtr));
+    }
+    if (payload != null) {
+      data.put(PAYLOAD_COL, payload);
+    }
+    return Bytes.toBytes(GSON.toJson(data));
```

Comment:
That's a bit heavy to use Gson. Maybe better to use encoding something like, 1st byte to tell whether it is transactional or not (0 vs 1). If it is transactional (bytes[0] == 1), then the following 8 bytes is the tx write pointer, following with the payload. Otherwise (bytes[0] == 0), the rest of the array is the payload.